### DISCUSSION TITLE
Banking App Business Use Case

### DISCUSSION THREAD

**Discussion by janfernandosmoothstack** [finos/traderX#183] [Ideas]
# Banking App Business Use Case

## 07/16/24: FINOS Tech Sprint 2024 Kick-off
[Watch the Smoothstack recording](https://www.finos.org/hubfs/FINOS%20events/2024-07%20Tech%20Sprint/Use%20Case%20Pitches/Tech%20Sprint%20Use%20Cases%20Smoothstack.mp4)
[Watch the Open Bank Project recording](https://www.finos.org/hubfs/FINOS%20events/2024-07%20Tech%20Sprint/Use%20Case%20Pitches/Tech%20Sprint%20Use%20Cases-%20Open%20Bank%20Tsobi.mp4)

### Business Problem
Customers need an integrated solution within their banking application to manage investments efficiently. They are currently using separate platforms to manage their bank accounts and investment portfolios, resulting in continuous transaction fees and an overall fragmented and time-consuming experience. 

### Proposed Solution
Integrate a trading feature within the banking application that will allow customers to manage their investments, providing a seamless experience for monitoring and executing trades. 

1. **Integration with traderX Data API** 

    Viewing & Monitoring Investment Opportunities: Integrate the traderX Data API to access real-time market data, such as: interest rates, yields, maturity dates, and historical performance data of available short-term financial instruments, enabling informed investment decisions. 

2. **Integration with traderX Trading API** 

    Trading: Integrate the traderX Trading API to facilitate the purchasing and selling of short-term financial instruments directly from the banking application, ensuring all transactions are at current market prices and are to and from the customers personal bank account; allowing users to manage their portfolio in real-time. 

#

### Business Problem
As a CFO, you need to ensure your company's excess cash is not sitting idle. You seek to generate returns by investing in short-term financial instruments, but manual analysis of market conditions and constant monitoring of investment opportunities is time-consuming and prone to human error.

### Proposed Solution: AI Agents

1. **Integration with traderX Data API:**

    - The AI Agent 1 connects to traderX's Data API to access real-time market data, including: Interest rates for various short-term instruments (e.g., Treasury bills, commercial paper, money market funds); Yields and maturity dates of available instruments; Historical performance data for risk assessment.

2. **AI-Driven Investment Strategy:**

    - The AI Agent 2 leverages advanced algorithms (e.g., reinforcement learning) to analyze the incoming market data.
    - It considers factors such as: Current and projected interest rates; Risk tolerance based on the company's financial profile; The need for liquidity (i.e., how quickly the funds might be needed for operations).
    - The agent develops an optimized investment strategy, determining the optimal allocation of funds across different short-term instruments.

3. **Automated Execution via traderX Trading API:**

    - The AI Agent 3 connects to traderX's Trading API to automatically execute trades based on the calculated strategy.
    - It can place buy and sell orders, monitor positions, and adjust the portfolio as market conditions change.

4. **Continuous Monitoring and Reporting:**

    - The AI Agent 4 continuously monitors the performance of the investments and generates reports for the CFO.
    - It provides insights into realized returns, projected returns, and potential risks

#
### High Level Architecture

![bank-architecture2](https://github.com/user-attachments/assets/18d9b7ba-767d-4cee-89cf-66cccc474dc1)

### Resources:
[Banking App Use Case - Summary.pdf](https://github.com/user-attachments/files/16146775/Banking.App.Use.Case.-.Summary.pdf)

[Banking App Use Case - Accounts - Investing - Customer Portal.pdf](https://github.com/user-attachments/files/16150791/Banking.App.Use.Case.-.Accounts.-.Investing.-.Customer.Portal.pdf)

[Banking App Use Case - All Features.pdf](https://github.com/user-attachments/files/16150778/Banking.App.Use.Case.-.All.Features.pdf)

**Comment by DovOps** [comment]
@janfernandosmoothstack  - Thank you for sharing these documents. I think that this is a nice set of use cases and requirements to build on top of / or in the style of TraderX. @maoo  I propose we use the next meeting to discuss.

**Comment by maoo** [comment]
Very interesting @janfernandosmoothstack ! I'd also suggest to mention this as a track for the TraderX event we're organizing. Any problem if I convert this to a discussion, in order to simplify the conversation with developers who may be interested to help out?

**Comment by janfernandosmoothstack** [comment]
@maoo No problem at all, please go ahead and convert to a discussion.

**Reply by maoo** [reply]
Thanks @janfernandosmoothstack ! I went through the documents and it's a very interesting idea, with extensive documentation.

We have just announced the FINOS Tech Sprint 2024, starting today and running until September 30 (OSFF New York), when developers will be able to showcase their work in front of our members and community; throughout the summer, we will organize in-person events for developers to meet, align and discuss; it's also an opportunity for FINOS Project Leads to deliver some talks around technologies involved. The first in-person event is on July 16th, New York.

You can find all the info you need and the registration form on [https://www.finos.org/open-source-in-finance-tech-sprint-2024](https://www.finos.org/open-source-in-finance-tech-sprint-2024)

It would be great if you could update the description of this discussion with:
- Business problem
- Proposed solution
- High level architecture
- Related technologies/platforms
- Roadmap and breakdown

This would help developers getting interested in this track and help out with the implementation.

If you have any questions, please reach out help@finos.org and we'd happy to assist. Thank you!

**Comment by lucaborella89** [comment]
@simonredfern this is very relevant for OBP, check it out!

**Comment by lucaborella89** [comment]
@janfernandosmoothstack I work with @maoo and got inspired by your comments. See below an example to update the description: 

**Challenge**:

As a CFO, you need to ensure your company's excess cash is not sitting idle. You seek to generate returns by investing in short-term financial instruments, but manual analysis of market conditions and constant monitoring of investment opportunities is time-consuming and prone to human error.

**Solution**: AI Agents

1. Integration with traderX Data API:

- The AI Agent 1 connects to traderX's Data API to access real-time market data, including: Interest rates for various short-term instruments (e.g., Treasury bills, commercial paper, money market funds); Yields and maturity dates of available instruments; Historical performance data for risk assessment.

2. AI-Driven Investment Strategy:

- The AI Agent 2 leverages advanced algorithms (e.g., reinforcement learning) to analyze the incoming market data.
- It considers factors such as: Current and projected interest rates; Risk tolerance based on the company's financial profile; The need for liquidity (i.e., how quickly the funds might be needed for operations).
- The agent develops an optimized investment strategy, determining the optimal allocation of funds across different short-term instruments.

3. Automated Execution via traderX Trading API:

- The AI Agent 3 connects to traderX's Trading API to automatically execute trades based on the calculated strategy.
- It can place buy and sell orders, monitor positions, and adjust the portfolio as market conditions change.

4. Continuous Monitoring and Reporting:

- The AI Agent 4 continuously monitors the performance of the investments and generates reports for the CFO.
- It provides insights into realized returns, projected returns, and potential risks

I hope this helps!

**Reply by janfernandosmoothstack** [reply]
@lucaborella89 Thank you for this example. It will help a lot as we update the description.

**Reply by janfernandosmoothstack** [reply]
@lucaborella89 I have updated the description using your example as inspiration. If you have a moment, would appreciate your thoughts on it.
Additionally, with your permission, I would like to include your challenge as kind of a Stage 2 challenge to our initial one in case a developer would like to take it to the next level.

**Reply by lucaborella89** [reply]
@janfernandosmoothstack thanks for updating it. It sounds good to me. You envision a sort of "[robinhood](https://robinhood.com/eu/en/about/crypto/)", i.e. a bank app that enables end users check prices and execute. Yes, "my" challenge could be a "stage 2" challenge (@aikarjal fyi)

**Reply by janfernandosmoothstack** [reply]
@lucaborella89 Yeah that was kind of the idea. The overall use case was designed to be a one-stop shop for a customers banking needs. I have added the example you shared. I hope the information is presented well.

**Reply by lucaborella89** [reply]
@janfernandosmoothstack would you have time to 10-min pitch this use case (1st challenge) on the 16th of July from 10am-12pm ET?

**Reply by janfernandosmoothstack** [reply]
@lucaborella89 yes I am available during that time. This will be done virtually right?
Would it be ok if I included one of the trainees in our program who has built this use case, just to give a visual of a banking app that needs those features added?

**Reply by lucaborella89** [reply]
@janfernandosmoothstack sure, and yes, it will be virtually. Please reach out luca.borella@finos.org so that I put you in touch with my colleagues

**Comment by lucaborella89** [comment]
@aikarjal, please check this use case, in particular 2nd Challenge.

**Comment by janfernandosmoothstack** [comment]
![Confluence Header (1)](https://github.com/user-attachments/assets/2ffbdec4-7455-4cbe-a773-ae117d3ac85e)

## Smoothstack Banking Application

[Final Presentation Recording](https://smoothstack0-my.sharepoint.com/:v:/g/personal/prashanth_ram_smoothstack_com/ET-_8DeMegBBnn5Pg8RVAaABB58NDZyjUQcJkOSdjQcPpg?e=2QtlHQ)

[Presentation Overview_ AI-Powered Investment Services via TraderX Integration.pdf](https://github.com/user-attachments/files/17935538/Presentation.Overview_.AI-Powered.Investment.Services.via.TraderX.Integration.pdf)

## TraderX Integration Use Cases

### Use Case 1: TraderX Trade Services Integration

#### Objective
Execute and process trades from the banking application.

#### Key Developments

- **Investment Orchestrator:** Dedicated database for tracking transactions, positions, portfolios, and investors.
- **Live Data Feed:** Real-time market data simulation with historical stock data.
- **Frontend Pages:** Enhanced user interface for improved interaction with trade services.

### Use Case 2: Stock Advice System

#### Objective
Enable users to request stock advice from a Large Language Model (LLM).

#### Solution

- **Data Aggregation:** Gather relevant market data upon user request.
- **LLM Interaction:** Send data to a trained LLM for structured advice.
- **Automated Trade Suggestions:** Provide users with informed trade order options to accept or dismiss

## High Level Architecture
![image](https://github.com/user-attachments/assets/1559d64c-eaee-476f-9c66-3b9eadd89c91)
